### ISSUE TITLE
kitty doesn't render ranger properly

### ISSUE THREAD

**Issue by katawful** [kovidgoyal/kitty#2490]
When using ranger with VCS enabled, VCS enabled folders are poorly rendered by kitty. I've tested this in a few other terminals: konsole, urxvt, st, and xterm. Besides xterm not rendering anything, every other terminal doesn't have this issue. [This](https://github.com/ranger/ranger/issues/1889) issue on ranger's git has my issues including pictures. I don't believe its a ranger issue and the fact that other terminals work fine with ranger makes me believe its an issue on kitty's end. I'm using a monospace font so it's probably something to do with how kitty handles this situation.

I'm using kitty 0.17.1.1 on Arch Linux

**Comment by kovidgoyal** [comment]
This will almost certainly be because ranger's wcwidth() and kitty's
wcwidth() disagree, for some characters used in the rendering. kitty's
is based on the latest unicode standard. Many older terminals and
presumably ranger too use the system wcwidth() so they agree. This is
broken (it breaks over ssh for instance on systems with different
wcwidths). Although from your screenshot I dont see any non-ascii
characters.

**Comment by toonn** [comment]
@kovidgoyal, I had katawful try a recent PR that changes ranger's `wcwidth` to use the one from the python stdlib `wcwidth` and they say it didn't fix the problem. What more can we do to fix our character width functions?

**Comment by kovidgoyal** [comment]
I couldn't tell from the screenshot which characters were involved. Can
you tell me, and I will check.

**Comment by toonn** [comment]
@katawful, could you replace the symbols in `ranger/gui/widgets/__init__.py` with ASCII characters? I think that's the easiest way to see which ones are causing trouble. It's probably either `✓` or `⌂`.

**Comment by kovidgoyal** [comment]
Both of those are not emoji and should therefore always have wcwidth = 1
You can check waht width kitty assigns a string, with 

kitty +runpy "from kitty.fast_data_types import *; import sys; print(wcswidth(sys.argv[-1]))" ⌂

**Comment by katawful** [comment]
Yeah, that returned 1

**Comment by toonn** [comment]
Ranger does to. @katawful, could you try pinning down which symbols are causing the non-indented borders?

**Comment by katawful** [comment]
I booted up ranger in urxvt with the same font I'm using (MesloLGS NF) and it seems like kitty isn't rendering a specific character. I'm not really sure what this character is exactly so I'm just going to upload a picture. I believe it appears for any untracked folder/file

![2020-03-29~15:14:13](https://user-images.githubusercontent.com/45222045/77858270-6f242f00-71d0-11ea-9c8a-d7bc892282cc.png)

kitty on left, uxvrt on right

**Comment by Luflosi** [comment]
Can you copy the character and paste it here?

**Comment by katawful** [comment]
I can't seem to copy it from urxvt, it just gives me "?", which it also gives for the ranger border character. The character seems to be for untracked files ignored by .gitignore. Untracked files that aren't added have "?" for their marker so it seems like urxvt has a hard time copying it

**Comment by trygveaa** [comment]
It's probably `·`, from https://github.com/ranger/ranger/blob/v1.9.3/ranger/gui/widgets/__init__.py#L23

**Comment by katawful** [comment]
I put that character into a file and read it with cat and it managed to render properly along with vim. Less doesn't render it, but not unexpected. However, ranger can't render it with its file preview function (reading the widgets/__init__.py file) still. urxvt manages to render that fine when previewing (though it misses one other character). I'm not sure where the problem even lies anymore, different fonts set in kitty don't seem to change the rendering issue

**Comment by trygveaa** [comment]
I tested now, and it renders fine in both less and ranger for me with kitty. I haven't tried with the font you're using though (I'm using JetBrains Mono).

**Comment by katawful** [comment]
I tried it with that font and it has the same issue with everything. I tried using less with the file with the character in urxvt and its not rendering either. Could I be missing something with my system? I installed kitty via the arch community repo and I assumed it grabbed any necessary dependencies

**Comment by kovidgoyal** [comment]
Try installing the symbola font. An use --debug-font-fallback to check which fonts are being used. Although I dont really see why not rendering a character would cause alignment issues, that can only happen because of wcwidth() mismatches.

**Comment by toonn** [comment]
Ranger's reported width for that dot is 1 as well. @trygveaa, does this mean you can't reproduce katawful's problem? I.e., `:set draw_borders both`, `:set vcs_aware true` and `:set vcs_backend_git local`, you don't get misaligned borders when you do this?

**Comment by trygveaa** [comment]
Yeah, it looks fine here.
![shot_2020-03-30_10-33-56_1739x1021](https://user-images.githubusercontent.com/601966/77892691-4608aa80-7273-11ea-8946-7bb7329cbc01.png)

**Comment by katawful** [comment]
So I went through ranger in multiple scenarios. Its using Nimbus Mono PS as the preloaded font face. I set kitty to use Symbola, and also tried terminus. I even tried using bash instead of zsh. The only real thing that --debug-font-fallback gave me was an unknown character after ESC when starting ranger:   

```
[090 12:03:04.705863] [PARSE ERROR] Unknown char after ESC: 0x6b   
[090 12:03:04.705889] [PARSE ERROR] Unknown char after ESC: 0x5c
```
Not really sure what else I can do, the rendering problem still persists. I've tried fresh config files for ranger and kitty with no change and all the various different fonts don't really change anything. Kitty *can* render the character in question fine when echoing or using cat, and neovim also has no issue. Nano can't though, giving me a similar problem to less so it seems like there's something missing on my system instead
